### ISSUE TITLE
DFU help

### ISSUE THREAD

**Issue by ert78gb** [UltimateHackingKeyboard/firmware#919]
Please someone help me in the device firmware upgrade. 

I read the https://www.usb.org/sites/default/files/DFU_1.1.pdf documentation.
I also created a very minimal [repo](https://github.com/ert78gb/dfu) to simulate the firmware upgrade procedure.
The repo uses https://github.com/node-usb/node-usb node package that implement the https://developer.mozilla.org/en-US/docs/Web/API/USBDevice/deviceProtocol 

if you
- clone the repo
- run `npm ci`
- press the reset button on the right side on the keyboard
- run `node usb-test.js`

It print uhk bootloader 
```javascript
const device = {
  device: {
    busNumber: 1,
    deviceAddress: 3,
    deviceDescriptor: {
      bLength: 18,
      bDescriptorType: 1,
      bcdUSB: 512,
      bDeviceClass: 239,
      bDeviceSubClass: 2,
      bDeviceProtocol: 1,
      bMaxPacketSize0: 64,
      idVendor: 7504,
      idProduct: 24869,
      bcdDevice: 260,
      iManufacturer: 1,
      iProduct: 2,
      iSerialNumber: 3,
      bNumConfigurations: 1
    },
    portNumbers: [ 1, 1 ],
    interfaces: undefined
  },
  configurations: [
    {
      configurationValue: 1,
      configurationName: '',
      interfaces: [
        {
          interfaceNumber: 0,
          alternate: {
            alternateSetting: 0,
            interfaceClass: 2,
            interfaceSubclass: 2,
            interfaceProtocol: 0,
            interfaceName: '',
            endpoints: [
              {
                endpointNumber: 1,
                direction: 'in',
                type: 'interrupt',
                packetSize: 16
              }
            ]
          },
          alternates: [
            {
              alternateSetting: 0,
              interfaceClass: 2,
              interfaceSubclass: 2,
              interfaceProtocol: 0,
              interfaceName: '',
              endpoints: [
                {
                  endpointNumber: 1,
                  direction: 'in',
                  type: 'interrupt',
                  packetSize: 16
                }
              ]
            }
          ],
          claimed: false
        },
        {
          interfaceNumber: 1,
          alternate: {
            alternateSetting: 0,
            interfaceClass: 10,
            interfaceSubclass: 0,
            interfaceProtocol: 0,
            interfaceName: '',
            endpoints: [
              {
                endpointNumber: 2,
                direction: 'in',
                type: 'bulk',
                packetSize: 64
              },
              {
                endpointNumber: 1,
                direction: 'out',
                type: 'bulk',
                packetSize: 64
              }
            ]
          },
          alternates: [
            {
              alternateSetting: 0,
              interfaceClass: 10,
              interfaceSubclass: 0,
              interfaceProtocol: 0,
              interfaceName: '',
              endpoints: [
                {
                  endpointNumber: 2,
                  direction: 'in',
                  type: 'bulk',
                  packetSize: 64
                },
                {
                  endpointNumber: 1,
                  direction: 'out',
                  type: 'bulk',
                  packetSize: 64
                }
              ]
            }
          ],
          claimed: false
        }
      ]
    }
  ],
  usbVersionMajor: 2,
  usbVersionMinor: 0,
  usbVersionSubminor: 0,
  deviceClass: 239,
  deviceSubclass: 2,
  deviceProtocol: 1,
  vendorId: 7504,
  productId: 24869,
  deviceVersionMajor: 1,
  deviceVersionMinor: 0,
  deviceVersionSubminor: 4,
  controlTransferAsync: [Function: bound controlTransfer],
  setConfigurationAsync: [Function: bound setConfiguration],
  resetAsync: [Function: bound reset],
  getStringDescriptorAsync: [Function: bound getStringDescriptor],
  manufacturerName: 'Ultimate Gadget Laboratories',
  productName: 'UHK 80 right half bootloader',
  serialNumber: '1BB6E7881E17A628'
}
```

if I understand well the reset button start the bootloader that means the USB descriptor should contain the DFU descriptor. The 15th page of the DFU pdf describe the 
<img width="832" alt="Screenshot 2024-07-15 at 21 52 04" src="https://github.com/user-attachments/assets/7b4cb607-a0e5-4a5a-b56f-6bb3d33faa2b">

I expect `interface` where `bInterfaceClass` is `0xfe` that `254` decimal. But I don't find it in the interface list.

Independently of the above I tried almost every command on every interface but I get `{ status: 'stall' }` response.

Could you help me in the protocol? Which usb command should send when on which interface?

**Comment by mondalaci** [comment]
@benedekkupper Would you please help with this?

**Comment by benedekkupper** [comment]
I worked with DFU before on STM32 and also did host side tool in C# back in the day, so I should be able to help. However the device info you printed here looks nothing like a DFU device, more like a USB CDC serial device. So I’d start by checking what’s going on in the firmware side. @mondalaci did you test the bootloader with any zephyr/nrf tool they provide for their FW?

**Comment by mondalaci** [comment]
@benedekkupper Based on `productName: 'UHK 80 right half bootloader'` alone, @ert78gb has managed to trigger the bootloader via the reset button.

For me, when entering the bootloader, `dmesg` shows:

```
[153369.164204] usb 1-9.2.1: USB disconnect, device number 112
[153370.114894] usb 1-9.2.1: new full-speed USB device number 113 using xhci_hcd
[153370.241711] usb 1-9.2.1: New USB device found, idVendor=1d50, idProduct=6125, bcdDevice= 1.04
[153370.241714] usb 1-9.2.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[153370.241716] usb 1-9.2.1: Product: UHK 80 right half bootloader
[153370.241717] usb 1-9.2.1: Manufacturer: Ultimate Gadget Laboratories
[153370.241717] usb 1-9.2.1: SerialNumber: 9ED07860CD85B5C0
[153370.251126] cdc_acm 1-9.2.1:1.0: ttyACM1: USB ACM device
```

`node usb-test.js` yields:

```
WebUSBDevice {
  device: {
    busNumber: 1,
    deviceAddress: 117,
    deviceDescriptor: {
      bLength: 18,
      bDescriptorType: 1,
      bcdUSB: 512,
      bDeviceClass: 239,
      bDeviceSubClass: 2,
      bDeviceProtocol: 1,
      bMaxPacketSize0: 64,
      idVendor: 7504,
      idProduct: 24869,
      bcdDevice: 260,
      iManufacturer: 1,
      iProduct: 2,
      iSerialNumber: 3,
      bNumConfigurations: 1
    },
    portNumbers: [ 9, 2, 1 ],
    interfaces: undefined
  },
  configurations: [
    {
      configurationValue: 1,
      configurationName: '',
      interfaces: [
        {
          interfaceNumber: 0,
          alternate: {
            alternateSetting: 0,
            interfaceClass: 2,
            interfaceSubclass: 2,
            interfaceProtocol: 0,
            interfaceName: '',
            endpoints: [
              {
                endpointNumber: 1,
                direction: 'in',
                type: 'interrupt',
                packetSize: 16
              }
            ]
          },
          alternates: [
            {
              alternateSetting: 0,
              interfaceClass: 2,
              interfaceSubclass: 2,
              interfaceProtocol: 0,
              interfaceName: '',
              endpoints: [
                {
                  endpointNumber: 1,
                  direction: 'in',
                  type: 'interrupt',
                  packetSize: 16
                }
              ]
            }
          ],
          claimed: false
        },
        {
          interfaceNumber: 1,
          alternate: {
            alternateSetting: 0,
            interfaceClass: 10,
            interfaceSubclass: 0,
            interfaceProtocol: 0,
            interfaceName: '',
            endpoints: [
              {
                endpointNumber: 2,
                direction: 'in',
                type: 'bulk',
                packetSize: 64
              },
              {
                endpointNumber: 1,
                direction: 'out',
                type: 'bulk',
                packetSize: 64
              }
            ]
          },
          alternates: [
            {
              alternateSetting: 0,
              interfaceClass: 10,
              interfaceSubclass: 0,
              interfaceProtocol: 0,
              interfaceName: '',
              endpoints: [
                {
                  endpointNumber: 2,
                  direction: 'in',
                  type: 'bulk',
                  packetSize: 64
                },
                {
                  endpointNumber: 1,
                  direction: 'out',
                  type: 'bulk',
                  packetSize: 64
                }
              ]
            }
          ],
          claimed: false
        }
      ]
    }
  ],
  usbVersionMajor: 2,
  usbVersionMinor: 0,
  usbVersionSubminor: 0,
  deviceClass: 239,
  deviceSubclass: 2,
  deviceProtocol: 1,
  vendorId: 7504,
  productId: 24869,
  deviceVersionMajor: 1,
  deviceVersionMinor: 0,
  deviceVersionSubminor: 4,
  controlTransferAsync: [Function: bound controlTransfer],
  setConfigurationAsync: [Function: bound setConfiguration],
  resetAsync: [Function: bound reset],
  getStringDescriptorAsync: [Function: bound getStringDescriptor],
  manufacturerName: 'Ultimate Gadget Laboratories',
  productName: 'UHK 80 right half bootloader',
  serialNumber: '9ED07860CD85B5C0'
}
/home/laci/download/x/dfu/node_modules/usb/dist/webusb/webusb-device.js:194
            throw new Error(`controlTransferIn error: ${error}`);
                  ^

Error: controlTransferIn error: Error: LIBUSB_ERROR_IO
    at WebUSBDevice.controlTransferIn (/home/laci/download/x/dfu/node_modules/usb/dist/webusb/webusb-device.js:194:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async file:///home/laci/download/x/dfu/usb-test.js:30:18

Node.js v20.9.0
```

**Comment by benedekkupper** [comment]
dmesg also says it’s not a DFU device but a CDC ACM serial one

**Comment by mondalaci** [comment]
@benedekkupper Whatever the case is, mcuboot gets triggered, and the firmware can be uploaded using:

`mcumgr --conntype serial --connstring "/dev/ttyACM4,baud=115200,mtu=4096" image upload -e app_update.bin`

**Comment by ert78gb** [comment]
I could use serial port if someone describe the protocol.

@benedekkupper I coded in c# ~6 years, but I it was 20 years ago :). You can share the c# code

**Comment by benedekkupper** [comment]
So the mcumgr is also talking over the serial interface, not over DFU… so either we stick with that and find the documentation for the specific serial protocol, or we check how to turn it into DFU protocol instead.

**Comment by mondalaci** [comment]
@benedekkupper Would you open a thread at https://devzone.nordicsemi.com/ and ask how to expose standard DFU instead of serial via MCUboot? If it's not possible to use standard DFU, we should ask them about the serial protocol specification used.

I could open a thread, but I'm not familiar with the difference between DFU vs serial, and you could elaborate on it in your question.

**Comment by ert78gb** [comment]
If Nordic has lightweight nodejs library it is also good.  :) The lightweight means contains only the communication feature without UI libraries and other tools . Because they have the https://github.com/NordicSemiconductor/pc-nrfconnect-shared but it is full with other things and expect to install the nordic sdk. Other nodejs packages are all deprecated, so I don't know they don't wanna maintain them or changed/deprecated the protocols that used by these libs.

**Comment by benedekkupper** [comment]
From what I can read from the Kconfig of mcuboot, this is what we need, but pay attention to the note, I'm not sure if we fit into this limitation:
```
config BOOT_USB_DFU_WAIT
	bool "Wait for a prescribed duration to see if USB DFU is invoked"
	help
	  If y, MCUboot waits for a prescribed duration of time to allow
	  for USB DFU to be invoked. Please note DFU always updates the
	  slot1 image.
```
and at the same time remove all serial related Kconfig options:
```
#CONFIG_BOOT_SERIAL_CDC_ACM=y
#CONFIG_BOOT_SERIAL_PIN_RESET=y
#CONFIG_BOOT_SERIAL_WAIT_FOR_DFU=y
#CONFIG_BOOT_SERIAL_WAIT_FOR_DFU_ALWAYS=n
#CONFIG_BOOT_SERIAL_WAIT_FOR_DFU_TIMEOUT=3000
#CONFIG_BOOT_SERIAL_NO_APPLICATION=y
#CONFIG_BOOT_SERIAL_MAX_RECEIVE_SIZE=4096
#CONFIG_BOOT_SERIAL_BOOT_MODE=y
```

**Comment by mondalaci** [comment]
@benedekkupper What BOOT_USB_DFU_WAIT is meant to solve?

We don't use this Kconfig option because it always waits on bootup. We use an alternative solution that only waits when the reset button is used. It's also possible to trigger the bootloader with a USB command.

**Comment by benedekkupper** [comment]
AFAICT we need to select one of the BOOT_USB_DFU* Kconfig options in order to get a DFU interface. That's either BOOT_USB_DFU_GPIO or BOOT_USB_DFU_WAIT, and I don't know (yet) how the GPIO option works.

**Comment by mondalaci** [comment]
@benedekkupper I've tried to add the suggested Kconfig options to mcuboot.conf of which only `CONFIG_BOOT_USB_DFU_WAIT=y` built successfully.

According to `dmesg`, MCUboot enumerated the same way both with and without this option:

```
[533249.989317] usb 1-9.2.1: new full-speed USB device number 11 using xhci_hcd
[533250.116877] usb 1-9.2.1: New USB device found, idVendor=1d50, idProduct=6125, bcdDevice= 1.04
[533250.116895] usb 1-9.2.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[533250.116902] usb 1-9.2.1: Product: UHK 80 right half bootloader
[533250.116907] usb 1-9.2.1: Manufacturer: Ultimate Gadget Laboratories
[533250.116911] usb 1-9.2.1: SerialNumber: 9ED07860CD85B5C0
[533250.127161] cdc_acm 1-9.2.1:1.0: ttyACM2: USB ACM device
```

Please look into how to make MCUboot expose DFU, assuming it's possible and matters. I'm also curious how to tell that DFU is exposed.

**Comment by mondalaci** [comment]
@benedekkupper Please make this a priority. We must finalize the bootloader so that our dongle supplier can flash it to the soon-to-be-manufactured dongles.

**Comment by benedekkupper** [comment]
If I modify the mcuboot.conf to this:
```python
CONFIG_PINCTRL=y
CONFIG_PM_PARTITION_SIZE_MCUBOOT=0x10000
CONFIG_LOG=y
CONFIG_SIZE_OPTIMIZATIONS=y
#CONFIG_SERIAL=y
CONFIG_UART_LINE_CTRL=y
#CONFIG_MCUBOOT_SERIAL=y
CONFIG_BOOT_USB_DFU_WAIT=y
#CONFIG_BOOT_SERIAL_CDC_ACM=y
#CONFIG_BOOT_SERIAL_PIN_RESET=y
#CONFIG_BOOT_SERIAL_WAIT_FOR_DFU=y
#CONFIG_BOOT_SERIAL_WAIT_FOR_DFU_ALWAYS=n
#CONFIG_BOOT_SERIAL_WAIT_FOR_DFU_TIMEOUT=3000
#CONFIG_BOOT_SERIAL_NO_APPLICATION=y
#CONFIG_BOOT_SERIAL_MAX_RECEIVE_SIZE=4096
#CONFIG_BOOT_SERIAL_BOOT_MODE=y
CONFIG_USB_DEVICE_VID=0x1D50
CONFIG_USB_DEVICE_PID=0x6125
CONFIG_USB_DEVICE_MANUFACTURER="Ultimate Gadget Laboratories"
CONFIG_DISABLE_FLASH_PATCH=y
# CONFIG_LOG_BACKEND_UART=y
CONFIG_LOG_BACKEND_RTT=n
# CONFIG_MCUBOOT_BOOTLOADER_MODE_SINGLE_APP=y
# CONFIG_MCUBOOT_SHELL=y
# CONFIG_MCUMGR_GRP_OS_ECHO=y
# CONFIG_MCUMGR_SMP_VERBOSE_ERR_RESPONSE=y
CONFIG_SINGLE_APPLICATION_SLOT=y
CONFIG_BOOT_MAX_LINE_INPUT_LEN=8192
CONFIG_RETAINED_MEM=y
CONFIG_RETENTION=y
CONFIG_RETENTION_BOOT_MODE=y
```
then I get a USB DFU interface on `lsusb -v -d 1d50:`.

**Comment by mondalaci** [comment]
Thanks! The mentioned changes indeed reveal the DFU descriptor. I'll use the following command from now on to check for its presence:

`lsusb -v -d 1d50: | grep 'Device Firmware Upgrade Interface Descriptor'`

Of all the Kconfig options you added/removed/changed, adding `CONFIG_BOOT_USB_DFU_WAIT=y` revealed the DFU interface, which I've just added to master with 4644d94f8a284b0cc63ad222393ad2d227f54f3c.

**Comment by ert78gb** [comment]
Think it is not working properly. Or need more info how to work with it.
I found the https://devanlai.github.io/webdfu/dfu-util/ tool. The source code https://github.com/devanlai/webdfu.

DFU is a standard protocol so I think the tool should work.

After the bootloader enumeration the USB device description is

```
WebUSBDevice {
  device: {
    busNumber: 1,
    deviceAddress: 4,
    deviceDescriptor: {
      bLength: 18,
      bDescriptorType: 1,
      bcdUSB: 512,
      bDeviceClass: 239,
      bDeviceSubClass: 2,
      bDeviceProtocol: 1,
      bMaxPacketSize0: 64,
      idVendor: 7504,
      idProduct: 24869,
      bcdDevice: 260,
      iManufacturer: 1,
      iProduct: 2,
      iSerialNumber: 3,
      bNumConfigurations: 1
    },
    portNumbers: [ 1, 1 ],
    interfaces: undefined
  },
  configurations: [
    {
      configurationValue: 1,
      configurationName: '',
      interfaces: [
        {
          interfaceNumber: 0,
          alternate: {
            alternateSetting: 0,
            interfaceClass: 2,
            interfaceSubclass: 2,
            interfaceProtocol: 0,
            interfaceName: '',
            endpoints: [
              {
                endpointNumber: 1,
                direction: 'in',
                type: 'interrupt',
                packetSize: 16
              }
            ]
          },
          alternates: [
            {
              alternateSetting: 0,
              interfaceClass: 2,
              interfaceSubclass: 2,
              interfaceProtocol: 0,
              interfaceName: '',
              endpoints: [
                {
                  endpointNumber: 1,
                  direction: 'in',
                  type: 'interrupt',
                  packetSize: 16
                }
              ]
            }
          ],
          claimed: false
        },
        {
          interfaceNumber: 1,
          alternate: {
            alternateSetting: 0,
            interfaceClass: 10,
            interfaceSubclass: 0,
            interfaceProtocol: 0,
            interfaceName: '',
            endpoints: [
              {
                endpointNumber: 2,
                direction: 'in',
                type: 'bulk',
                packetSize: 64
              },
              {
                endpointNumber: 1,
                direction: 'out',
                type: 'bulk',
                packetSize: 64
              }
            ]
          },
          alternates: [
            {
              alternateSetting: 0,
              interfaceClass: 10,
              interfaceSubclass: 0,
              interfaceProtocol: 0,
              interfaceName: '',
              endpoints: [
                {
                  endpointNumber: 2,
                  direction: 'in',
                  type: 'bulk',
                  packetSize: 64
                },
                {
                  endpointNumber: 1,
                  direction: 'out',
                  type: 'bulk',
                  packetSize: 64
                }
              ]
            }
          ],
          claimed: false
        }
      ]
    }
  ],
  usbVersionMajor: 2,
  usbVersionMinor: 0,
  usbVersionSubminor: 0,
  deviceClass: 239,
  deviceSubclass: 2,
  deviceProtocol: 1,
  vendorId: 7504,
  productId: 24869,
  deviceVersionMajor: 1,
  deviceVersionMinor: 0,
  deviceVersionSubminor: 4,
  controlTransferAsync: [Function: bound controlTransfer],
  setConfigurationAsync: [Function: bound setConfiguration],
  resetAsync: [Function: bound reset],
  getStringDescriptorAsync: [Function: bound getStringDescriptor],
  manufacturerName: 'Ultimate Gadget Laboratories',
  productName: 'UHK 80 right half bootloader',
  serialNumber: '1BB6E7881E17A628'
}
```

As the DFU documentation and https://github.com/devanlai/webdfu/blob/56d5d1d961587aca381a1ce97acbc8ff5d807acd/dfu-util/dfu.js#L39 shows the `bInterfaceClass` or `interfaceClass` field should be `0xfe` that `254` decimal, but I got decimal `10` and `2`.

Are you sure uhk implement properly the DFU ? If yes please give me some hints which USB command should I run.

**Comment by benedekkupper** [comment]
What you are seeing is still the serial interface, so either you have an earlier FW, or the kconfig still has the usb serial option in it, which overrides the DFU option.

I don’t know why there is no kconfig warning added here that would block enabling both, that’s something to complain about at zephyr’s side.

**Comment by ert78gb** [comment]
I downloaded the `firmware.bin` of the uhk-80 instead of the `merged.hex`. After I used the `merged.hex` the `interfaceClass: 0xfe` appeared so I can continue the develoment.

**Comment by ert78gb** [comment]
I use the reset button on the keyboard to start the bootloader. I do it to create a small prototype script that works properly and after I will migrate it to the agent ecosystem. Is it should work or not?

I can connect to the device and I can query the DFU state. It returns with `0` as `appIDLE` state. 
If I understand well this state is the normal application state = UHK firmware is operation.
Because the firmware indicate the DFU mode my expectation is the DFU state should be `dfuIDLE`, am I right?

If I start the DOWNLOAD operation I got error.

I recognised when I query the interface descriptor then the `bInterfaceProtocol: 1` but the documentation says it should be `2` 
<img width="1723" alt="Screenshot 2024-08-02 at 19 20 23" src="https://github.com/user-attachments/assets/72f4852c-ef96-4ae3-a6f6-00d11c206fdf">

What command should I run to move the bootloader from `appIDLE` to `dfuIDLE`?
